include/CLI/Validators.hpp

```diff
@@ -240,6 +264,137 @@ struct Range : public Validator {
+        using item_t = typename std::pointer_traits<T>::element_type::value_type;
+
+        // Make a local copy of the filter function, using a std::function if not one already
+        std::function<item_t(item_t)> filter_fn = filter_function;
+
+        // This is the type name for help, it will take the current version of the set contents
+        tname_function = [set]() {
+            std::stringstream out;
+            out << detail::type_name<item_t>() << " in {" << detail::join(*set, ",") << "}";
+            return out.str();
+        };
+
+        // This is the function that validates
+        // It stores a copy of the set pointer-like, so shared_ptr will stay alive
+        func = [set, filter_fn](std::string &input) {
+            for(const item_t &v : *set) {
+                item_t a = v;
+                item_t b;
+                if(!detail::lexical_cast(input, b))
+                    throw ConversionError(input); // name is added later
+
+                // The filter function might be empty, so don't filter if it is.
+                if(filter_fn) {
+                    a = filter_fn(a);
+                    b = filter_fn(b);
+                }
+
+                if(a == b) {
+                    // Make sure the version in the input string is identical to the one in the set
+                    // Requires std::stringstream << be supported on T.
+                    if(filter_fn) {
+                        std::stringstream out;
+                        out << v;
+                        input = out.str();
+                    }
+
+                    // Return empty error string (success)
+                    return std::string();
+                }
+            }
+
+            // If you reach this point, the result was not found
+            return input + " not in {" + detail::join(*set, ",") + "}";
+        };
+    }
+
+    /// This checks to see if an item is in a set: copy version.
+    template <typename T, typename F, enable_if_t<!is_copyable_ptr<T>::value, detail::enabler> = detail::dummy>
+    explicit IsMember(T set, F filter_function) {
+        // Get the type of the contained item - requires a container have ::value_type
+        using item_t = typename T::value_type;
+
+        // Make a local copy of the filter function, using a std::function if not one already
+        std::function<item_t(item_t)> filter_fn = filter_function;
+
+        // This is the type name for help, since the set contents can't change, we just capture this
+        std::stringstream out;
+        out << detail::type_name<item_t>() << " in {" << detail::join(set, ",") << "}";
+        tname = out.str();
+
+        // This is the function that validates
+        func = [set, filter_fn](std::string &input) {
+            for(const item_t &v : set) {
+                item_t a = v;
+                item_t b;
+                if(!detail::lexical_cast(input, b))
+                    throw ConversionError(input); // name is added later
```

Comment:
Should this be a validationError rather than ConversionError, since it will occur in the validation context, and having one of those throw a conversion Error when someone is checking for a Validation Error could be confusing,  possibly making no difference whatsoever but if you can't convert to the appropriate type of a given set it is a conversionError that generates a validationError.  Which one should be set to the user?

Reply:
The validator always runs first (that gives it the ability to affect the parse). So the following gives a conversion error:

```cpp
int myint;
app.add_option("--value", myint);
// Run with --value one
```

I didn't want the conversion error to change to a validation error here:

```cpp
int myint;
std::set<int> set = {1,2,3};
app.add_option("--value", myint)->check(CLI::IsMember(set));
// Run with --value one
```

But, both are true, and just saying "ValidationErrors" are stronger is probably fine.

You also have the odd mixed types that shouldn't happen but could:

```cpp
std::string mystr;
std::set<int> set = {1,2,3};
app.add_option("--value", mystr)->check(CLI::IsMember(set));
// Run with --value one
```
This fails only when checking the set and a ValidationError would fit better here, so I think I've convinced myself that ValidationError is better.